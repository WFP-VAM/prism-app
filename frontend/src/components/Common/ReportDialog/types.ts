import { Style } from '@react-pdf/types';
import { TFunctionResult } from 'i18next';

export interface PDFLegendDefinition {
  value: string | number | TFunctionResult;
  style: Style | Style[];
}

// This numbers depends on table header and row height
export const MAX_TABLE_ROWS_PER_PAGE = 29;
export const FIRST_PAGE_TABLE_ROWS = 9;
