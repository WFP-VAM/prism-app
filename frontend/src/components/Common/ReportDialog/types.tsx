export enum ReportType {
  Storm,
  Flood,
}
