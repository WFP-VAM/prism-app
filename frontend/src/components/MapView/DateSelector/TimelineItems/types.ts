import {
  CycloneName,
  ShortDate,
  TimeAndState,
} from 'context/anticipatoryAction/AAStormStateSlice/types';

export type DateItemStyle = {
  class: string;
  color: string;
  layerDirectionClass?: string;
  emphasis?: string;
};

export type DateJSON = Record<ShortDate, Record<CycloneName, TimeAndState[]>>;

export type WindStateReport = {
  states: TimeAndState[];
  cycloneName: string | null;
};
