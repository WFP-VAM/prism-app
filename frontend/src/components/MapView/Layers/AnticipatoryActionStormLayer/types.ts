// Q - still needed or needs to be improved?
import { FeatureCollection, Point, Feature } from 'geojson';

export enum FeaturePropertyDataType {
  analysis = 'analysis',
  forecast = 'forecast',
}

interface FeatureProperty {
  data_type: FeaturePropertyDataType;
  time: string;
  development: string;
  // maximum_wind_speed: number;
  // maximum_wind_gust: number;
  // wind_buffer_48: any;
  // wind_buffer_64: any;
}

export interface TimeSeries extends FeatureCollection<Point, FeatureProperty> {}
export interface TimeSeriesFeature extends Feature<Point, FeatureProperty> {}
