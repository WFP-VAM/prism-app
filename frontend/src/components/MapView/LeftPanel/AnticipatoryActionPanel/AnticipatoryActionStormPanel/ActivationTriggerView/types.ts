import { AACategory } from 'context/anticipatoryAction/AAStormStateSlice/parsedStormDataTypes';

export const AAPanelCategories: AACategory[] = [
  AACategory.Severe,
  AACategory.Moderate,
];

export const AADisplayCategory: {
  [key in AACategory]?: string;
} = {
  [AACategory.Severe]: ' >118 KM/H',
  [AACategory.Moderate]: ' >89 KM/H',
};
