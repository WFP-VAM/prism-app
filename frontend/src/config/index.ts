import { has, get } from 'lodash';
import { PublicClientApplication } from '@azure/msal-browser';
import afghanistan from './afghanistan';
import cambodia from './cambodia';
import cameroon from './cameroon';
import colombia from './colombia';
import cuba from './cuba';
import ecuador from './ecuador';
import global from './global';
import indonesia from './indonesia';
import jordan from './jordan';
import kyrgyzstan from './kyrgyzstan';
import mongolia from './mongolia';
import mozambique from './mozambique';
import myanmar from './myanmar';
import namibia from './namibia';
import nigeria from './nigeria';
import rbd from './rbd';
import sierraleone from './sierraleone';
import southsudan from './southsudan';
import srilanka from './srilanka';
import tajikistan from './tajikistan';
import tanzania from './tanzania';
import ukraine from './ukraine';
import zimbabwe from './zimbabwe';

// Upload the boundary URL to S3 to enable the use of the API in a local environment.
const DEFAULT_BOUNDARIES_FOLDER =
  'https://prism-admin-boundaries.s3.us-east-2.amazonaws.com';

const configMap = {
  afghanistan,
  cambodia,
  cameroon,
  colombia,
  cuba,
  ecuador,
  global,
  indonesia,
  jordan,
  kyrgyzstan,
  mongolia,
  mozambique,
  myanmar,
  namibia,
  nigeria,
  rbd,
  sierraleone,
  southsudan,
  srilanka,
  tajikistan,
  tanzania,
  ukraine,
  zimbabwe,
} as const;

type Country = keyof typeof configMap;

const DEFAULT: Country = 'mozambique';

const {
  REACT_APP_COUNTRY: COUNTRY,
  REACT_APP_OAUTH_CLIENT_ID: CLIENT_ID,
  REACT_APP_OAUTH_AUTHORITY: AUTHORITY,
  REACT_APP_OAUTH_REDIRECT_URI: REDIRECT_URI,
  REACT_APP_TESTING: TESTING,
} = process.env;

const safeCountry =
  COUNTRY && has(configMap, COUNTRY.toLocaleLowerCase())
    ? (COUNTRY.toLocaleLowerCase() as Country)
    : DEFAULT;

const {
  appConfig,
  defaultBoundariesFile,
  rawLayers,
  rawTables,
  rawReports,
}: {
  appConfig: Record<string, any>;
  defaultBoundariesFile: string;
  rawLayers: Record<string, any>;
  rawTables: Record<string, any>;
  rawReports: Record<string, any>;
} = configMap[safeCountry];

const translation = get(configMap[safeCountry], 'translation', {});

const msalConfig = {
  auth: {
    clientId: CLIENT_ID!,
    authority: AUTHORITY!,
    redirectUri: REDIRECT_URI!,
  },
};

const msalRequest = {
  scopes: ['openid', 'profile'],
};

const msalInstance = new PublicClientApplication(msalConfig);

const authRequired: boolean =
  !TESTING && get(appConfig, 'WFPAuthRequired', false);

const enableNavigationDropdown: boolean = get(
  appConfig,
  'enableNavigationDropdown',
  false,
);

const defaultBoundariesPath = `${DEFAULT_BOUNDARIES_FOLDER}/${defaultBoundariesFile}`;

export {
  appConfig,
  authRequired,
  safeCountry,
  defaultBoundariesPath,
  rawLayers,
  rawTables,
  rawReports,
  msalInstance,
  msalRequest,
  enableNavigationDropdown,
  translation,
};
