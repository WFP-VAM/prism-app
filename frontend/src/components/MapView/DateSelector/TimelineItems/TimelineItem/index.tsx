import { WithStyles, createStyles, withStyles } from '@material-ui/core';
import React, { memo } from 'react';
import 'react-datepicker/dist/react-datepicker.css';
import { DateItem, DateRangeType } from 'config/types';
import { binaryFind } from 'utils/date-utils';

const TimelineItem = memo(
  ({
    classes,
    concatenatedLayers,
    currentDate,
    dateItemStyling,
  }: TimelineItemProps) => {
    const hasNextItemDirectionForward = (
      matchingDate: DateItem,
      layerDates: DateItem[],
    ): boolean => {
      return (
        layerDates.indexOf(matchingDate) !== 0 &&
        !!layerDates[layerDates.indexOf(matchingDate) - 1].isStartDate
      );
    };

    const hasNextItemDirectionBackward = (
      matchingDate: DateItem,
      layerDates: DateItem[],
    ): boolean => {
      return (
        layerDates.indexOf(matchingDate) !== layerDates.length - 1 &&
        !!layerDates[layerDates.indexOf(matchingDate) + 1].isEndDate
      );
    };

    const isStartOrEndDate = (date: DateItem): boolean => {
      return !!date.isEndDate || !!date.isStartDate;
    };

    return (
      <>
        {concatenatedLayers.map(
          (layerDates: DateItem[], layerIndex: number) => {
            const idx = binaryFind<DateItem>(
              layerDates,
              new Date(currentDate.value).setHours(0, 0, 0, 0),
              (i: DateItem) => new Date(i.displayDate).setHours(0, 0, 0, 0),
            );
            const matchingDateItemInLayer: DateItem | undefined =
              idx > -1 ? layerDates[idx] : undefined;

            if (!matchingDateItemInLayer) {
              return null;
            }

            return (
              <React.Fragment key={Math.random()}>
                {/* Add a directional arrow forward if previous item is a start date */}
                {hasNextItemDirectionForward(
                  matchingDateItemInLayer,
                  layerDates,
                ) && (
                  <div
                    className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase}`}
                  />
                )}

                {/* Add a directional arrow backward if next item is an end date */}
                {hasNextItemDirectionBackward(
                  matchingDateItemInLayer,
                  layerDates,
                ) && (
                  <div
                    className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase} ${classes.layerDirectionBackwardBase}`}
                  />
                )}

                {/* Add a bold square if start or end date (emphasis), normal otherwise */}
                <div
                  className={`${
                    isStartOrEndDate(matchingDateItemInLayer)
                      ? dateItemStyling[layerIndex].emphasis
                      : dateItemStyling[layerIndex].class
                  }`}
                  role="presentation"
                />
              </React.Fragment>
            );
          },
        )}
      </>
    );
  },
);

const styles = () =>
  createStyles({
    layerDirectionBase: {
      display: 'block',
      position: 'absolute',
      borderTop: '5px solid transparent',
      borderBottom: '5px solid transparent',
      height: '0px',
      zIndex: 1,
      left: 0,
    },

    layerDirectionBackwardBase: {
      right: 0,
      transform: 'rotate(180deg)',
    },
  });

export interface TimelineItemProps extends WithStyles<typeof styles> {
  concatenatedLayers: DateItem[][];
  currentDate: DateRangeType;
  dateItemStyling: {
    class: string;
    color: string;
    layerDirectionClass?: string;
    emphasis?: string;
  }[];
}

export default withStyles(styles)(TimelineItem);
