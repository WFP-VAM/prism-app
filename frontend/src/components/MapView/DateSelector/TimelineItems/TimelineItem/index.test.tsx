import React from 'react';
import { render } from '@testing-library/react';
import { DateItem } from 'config/types';
import TimelineItem, { TimelineItemProps } from '.';

test('TimelineItem renders as expected', () => {
  // Arrange
  const currentDateTime = new Date().getTime();
  const props: Omit<TimelineItemProps, 'classes'> = {
    clickDate: () => {},
    concatenatedLayers: [
      [{ displayDate: currentDateTime, queryDate: currentDateTime }],
      [
        {
          displayDate: currentDateTime,
          queryDate: currentDateTime,
        },
        {
          displayDate: currentDateTime,
          queryDate: currentDateTime,
          isStartDate: true,
        },
      ],
    ],
    currentDate: {
      value: currentDateTime,
      label: '',
      month: '',
      isFirstDay: false,
      date: '',
    },
    index: 1,
    dateItemStyling: [
      { class: 'intersectionDate', color: 'White' },
      {
        class: 'layerOneDate',
        color: '',
        layerDirectionClass: 'layerDirectionClass',
      },
    ],
  };

  // Act
  const { container } = render(<TimelineItem {...props} />);

  // Assert
  expect(container).toMatchSnapshot();
});

test('TimelineItem renders as expected with data point starting after Timeline start date', () => {
  // Arrange
  const currentDateTime = 1640991600000; // 01-01-2022
  const firstLayerDataPoint = 1669896000000; // 01-12-2022
  const secondLayerDataPoint = 1670756400000; // 12-12-2022

  const layerDates: DateItem[] = [
    {
      displayDate: firstLayerDataPoint,
      queryDate: firstLayerDataPoint,
    },
    {
      displayDate: secondLayerDataPoint,
      queryDate: secondLayerDataPoint,
    },
  ];

  const props: Omit<TimelineItemProps, 'classes'> = {
    clickDate: () => {},
    concatenatedLayers: [layerDates, layerDates],
    currentDate: {
      value: currentDateTime,
      label: '',
      month: '',
      isFirstDay: false,
      date: '',
    },
    index: 1,
    dateItemStyling: [
      { class: 'intersectionDate', color: 'White' },
      {
        class: 'layerOneDate',
        color: '',
        layerDirectionClass: 'layerDirectionClass',
      },
    ],
  };

  // Act
  const { container } = render(<TimelineItem {...props} />);

  // Assert
  expect(container).toMatchSnapshot();
});
