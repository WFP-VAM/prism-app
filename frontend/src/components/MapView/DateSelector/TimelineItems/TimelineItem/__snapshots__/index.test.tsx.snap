// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItem renders as expected 1`] = `
<div>
  <div
    class="intersectionDate"
    role="presentation"
  />
  <div
    class="layerOneDate"
    role="presentation"
  />
</div>
`;

exports[`TimelineItem renders as expected with data point starting after Timeline start date 1`] = `<div />`;
